<regex>: Error C2664 in std::regex_traits::transform

**Describe the bug**
`std::regex_traits::transform` tries to use `const _Elem*` as internal iterator instead of user-defined iterator.

**Command-line test case**
```
d:\Temp2>type repro.cpp
#include <regex>
#include <iostream>

template <class It>
class forward_iterator
{
    It it_;

    template <class U> friend class forward_iterator;
public:
    typedef          std::forward_iterator_tag                 iterator_category;
    typedef typename std::iterator_traits<It>::value_type      value_type;
    typedef typename std::iterator_traits<It>::difference_type difference_type;
    typedef It                                                 pointer;
    typedef typename std::iterator_traits<It>::reference       reference;

    It base() const { return it_; }

    forward_iterator() : it_() {}
    explicit forward_iterator(It it) : it_(it) {}
    template <class U>
    forward_iterator(const forward_iterator<U>& u) : it_(u.it_) {}

    reference operator*() const { return *it_; }
    pointer operator->() const { return it_; }

    forward_iterator& operator++() { ++it_; return *this; }
    forward_iterator operator++(int)
    {
        forward_iterator tmp(*this); ++(*this); return tmp;
    }

    friend bool operator==(const forward_iterator& x, const forward_iterator& y)
    {
        return x.it_ == y.it_;
    }
    friend bool operator!=(const forward_iterator& x, const forward_iterator& y)
    {
        return !(x == y);
    }

    template <class T>
    void operator,(T const &) = delete;
};

int main()
{
    std::regex_traits<char> t;
    const char a[] = "a";
    const char B[] = "B";
    typedef forward_iterator<const char*> F;
    bool r = t.transform(F(a), F(a + 1)) > t.transform(F(B), F(B + 1));
    (void) r;
}

d:\Temp2>cl /EHsc /W4 .\repro.cpp
Microsoft (R) C/C++ Optimizing Compiler Version 19.27.29009.1 for x86
Copyright (C) Microsoft Corporation.  All rights reserved.

repro.cpp
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\regex(283): error C2664: 'std::basic_string<char,std::char_traits<char>,std::allocator<char>> std::collate<_Elem>::transform(const _Elem *,const _Elem *) const': cannot convert argument 1 from '_FwdIt' to 'const _Elem *'
        with
        [
            _Elem=char
        ]
        and
        [
            _FwdIt=F
        ]
        and
        [
            _Elem=char
        ]
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\regex(283): note: No user-defined-conversion operator available that can perform this conversion, or the operator cannot be called
c:\Program Files (x86)\Microsoft Visual Studio\2019\Preview\VC\Tools\MSVC\14.27.29009\include\locale(41): note: see declaration of 'std::collate<_Elem>::transform'
        with
        [
            _Elem=char
        ]
.\repro.cpp(52): note: see reference to function template instantiation 'std::basic_string<char,std::char_traits<char>,std::allocator<char>> std::_Regex_traits<char>::transform<F>(_FwdIt,_FwdIt) const' being compiled
        with
        [
            _FwdIt=F
        ]
.\repro.cpp(52): note: see reference to function template instantiation 'std::basic_string<char,std::char_traits<char>,std::allocator<char>> std::_Regex_traits<char>::transform<F>(_FwdIt,_FwdIt) const' being compiled
        with
        [
            _FwdIt=F
        ]
```

**Expected behavior**
The program should compile

**STL version**
```
Microsoft Visual Studio Professional 2019 Preview
Version 16.7.0 Preview 3.1
```

**Additional context**
The following is the equivalent of the repro:
https://github.com/microsoft/STL/blob/5be7d49c243979231dff35919d3a3d813d75d714/tests/libcxx/expected_results.txt#L564-L565

Also tracked by DevCom-1076807 and Microsoft-internal VSO-1142061 / [AB#1142061](https://devdiv.visualstudio.com/0bdbc590-a062-4c3f-b0f6-9383f67865ee/_workitems/edit/1142061).

--- a/stl/inc/regex
+++ b/stl/inc/regex
@@ -280,7 +280,7 @@ public:
 
     template <class _FwdIt>
     string_type transform(_FwdIt _First, _FwdIt _Last) const { // apply locale-specific transformation
-        string_type _Str(_First, _Last);
+        const string_type _Str(_First, _Last);
         return _Getcoll()->transform(_Str.data(), _Str.data() + _Str.size());
     }
 

--- a/stl/inc/regex
+++ b/stl/inc/regex
@@ -280,7 +280,8 @@ public:
 
     template <class _FwdIt>
     string_type transform(_FwdIt _First, _FwdIt _Last) const { // apply locale-specific transformation
-        return _Getcoll()->transform(_First, _Last);
+        string_type str(_First, _Last);
+        return _Getcoll()->transform(str.data(), str.data() + str.length());
     }
 
     template <class _FwdIt>

--- a/stl/inc/regex
+++ b/stl/inc/regex
@@ -271,7 +271,7 @@ public:
 
     _Elem translate(_Elem _Ch) const { // provide locale-sensitive mapping
         string_type _Res = _Getcoll()->transform(_STD addressof(_Ch), _STD addressof(_Ch) + 1);
-        return _Res.length() == 1 ? _Res[0] : _Ch;
+        return _Res.size() == 1 ? _Res[0] : _Ch;
     }
 
     _Elem translate_nocase(_Elem _Ch) const { // provide case-insensitive mapping
@@ -281,7 +281,7 @@ public:
     template <class _FwdIt>
     string_type transform(_FwdIt _First, _FwdIt _Last) const { // apply locale-specific transformation
         string_type _Str(_First, _Last);
-        return _Getcoll()->transform(_Str.data(), _Str.data() + _Str.length());
+        return _Getcoll()->transform(_Str.data(), _Str.data() + _Str.size());
     }
 
     template <class _FwdIt>

--- a/stl/inc/regex
+++ b/stl/inc/regex
@@ -280,8 +280,8 @@ public:
 
     template <class _FwdIt>
     string_type transform(_FwdIt _First, _FwdIt _Last) const { // apply locale-specific transformation
-        string_type str(_First, _Last);
-        return _Getcoll()->transform(str.data(), str.data() + str.length());
+        string_type _Str(_First, _Last);
+        return _Getcoll()->transform(_Str.data(), _Str.data() + _Str.length());
     }
 
     template <class _FwdIt>

--- a/tests/libcxx/expected_results.txt
+++ b/tests/libcxx/expected_results.txt
@@ -570,9 +570,6 @@ std/localization/locale.categories/category.monetary/locale.moneypunct/money_bas
 std/input.output/file.streams/fstreams/filebuf.virtuals/overflow.pass.cpp FAIL
 std/input.output/file.streams/fstreams/filebuf.virtuals/underflow.pass.cpp FAIL
 
-# GH-1004 <regex>: Error C2664 in std::regex_traits::transform
-std/re/re.traits/transform.pass.cpp FAIL
-
 # GH-1295 <array>: array<const T, 0> allows fill() and swap()
 std/containers/sequences/array/array.fill/fill.fail.cpp FAIL
 std/containers/sequences/array/array.swap/swap.fail.cpp FAIL

--- a/tests/libcxx/skipped_tests.txt
+++ b/tests/libcxx/skipped_tests.txt
@@ -570,9 +570,6 @@ localization\locale.categories\category.monetary\locale.moneypunct\money_base.pa
 input.output\file.streams\fstreams\filebuf.virtuals\overflow.pass.cpp
 input.output\file.streams\fstreams\filebuf.virtuals\underflow.pass.cpp
 
-# GH-1004 <regex>: Error C2664 in std::regex_traits::transform
-re\re.traits\transform.pass.cpp
-
 # GH-1295 <array>: array<const T, 0> allows fill() and swap()
 containers\sequences\array\array.fill\fill.fail.cpp
 containers\sequences\array\array.swap\swap.fail.cpp

